rlog/2024-06-24-gsub-largemsg-improvements.mdx

```diff
@@ -0,0 +1,430 @@
+For $D$ peers, the transmission time increases to $\tau_D^{1MB} = (80 \times D) + 100ms$, 
+triggering additional queuing delays (proportional to the transmission queue size) during each transmission round.  
+
+2. In practice, $\tau_1^{1MB}$ sometimes rises to several hundred milliseconds, 
+further exaggerating the abovementioned queuing delays.
+This rise is because TCP congestion avoidance algorithms usually limit maximum in-flight bytes in a round trip time (RTT) 
+based on the congestion window ($C_{wnd}$) and maximum segment size (MSS) to approximately ${C_{wnd} \times MSS}$, 
+with $C_{wnd}$ increasing with the data transfer for each flow. 
+Consequently, sending the same message through a newly established (cold) connection takes longer. 
+The message transfer time lowers as the $C_{wnd}$ grows. 
+Therefore, performance-improvement practices such as floodpublish, frequent mesh adjustment, and lazy sending 
+typically result in longer than expected message dissemination times for large messages (due to cold connections). 
+It is also worth mentioning that some TCP variants reset their $C_{wnd}$ after different periods of inactivity.
+
+3. Theoretically, the message transmission time to D peers $(\tau_D)$ remains the same 
+even if the message is relayed sequentially to all peers or simultaneous transmissions are carried out, 
+i.e., $\tau_D = \sum_{i=1}^{D} \tau_i$
+However, sequential transmissions finish early for individual peers, allowing them to relay early.
+It may result in quicker network-wide message dissemination. 
+
+4. A realistic network comprises nodes with dissimilar capabilities (bandwidth, link latency, compute, etc.).
+As the message disseminates, it's not uncommon for some peers to receive it much earlier than others. 
+Early gossip (IHAVE announcements) may bring in many IWANT requests to the early receivers (even from peers already receiving the same message), 
+which adds to their workload.  
+
+5. A busy peer (with a sizeable outgoing message queue) will enqueue (or simultaneously transfer) newly scheduled outgoing messages. 
+As a result, already scheduled messages are prioritized over peers' locally published messages, 
+introducing a significant initial delay to the locally published messages. 
+Enqueuing IWANT replies to the outgoing message queue can further exaggerate the problem. 
+The lack of adaptiveness and standardization in outgoing message prioritization are key factors that can lead to noticeable inconsistency 
+in message dissemination latency at each hop, even in similar network conditions.
+
+6. Message size directly contributes to peers' workloads in terms of processing and transmission time.
+It also raises the probability of simultaneous redundant transmissions to the same peer, 
+resulting in bandwidth wastage, congestion, and slow message propagation to the network. 
+Moreover, the benefits of sequential message relaying can be compromised by prioritizing slow (or busy) peer(s).

```

Comment:
unclear to me why we have the (s) after peer. I would omit the brackets.

Reply:
Sometimes we may choose a small group of peers